tensorpipe/test/channel/channel_test.h

```diff
@@ -18,49 +19,81 @@
 
 #include <gtest/gtest.h>
 
+#include <tensorpipe/test/peer_group.h>
+
 class ChannelTestHelper {
  public:
+  virtual ~ChannelTestHelper() = default;
+
   virtual std::shared_ptr<tensorpipe::channel::Context> makeContext(
       std::string id) = 0;
 
-  virtual ~ChannelTestHelper() = default;
+  virtual std::shared_ptr<PeerGroup> makePeerGroup() {
+    return std::make_shared<ThreadPeerGroup>();
+  }
 };
 
 class ChannelTest : public ::testing::TestWithParam<ChannelTestHelper*> {
+ protected:
+  std::shared_ptr<PeerGroup> peers_;
+
  public:
-  void testConnectionPair(
+  ChannelTest() : peers_(GetParam()->makePeerGroup()) {}
+
+  void testConnection(
       std::function<void(std::shared_ptr<tensorpipe::transport::Connection>)>
-          f1,
+          server,
       std::function<void(std::shared_ptr<tensorpipe::transport::Connection>)>
-          f2) {
-    auto context = std::make_shared<tensorpipe::transport::uv::Context>();
-    context->setId("harness");
+          client) {
     auto addr = "127.0.0.1";
 
-    {
-      tensorpipe::Queue<std::shared_ptr<tensorpipe::transport::Connection>> q1,
-          q2;
-
-      // Listening side.
-      auto listener = context->listen(addr);
-      listener->accept(
-          [&](const tensorpipe::Error& error,
-              std::shared_ptr<tensorpipe::transport::Connection> connection) {
-            ASSERT_FALSE(error) << error.what();
-            q1.push(std::move(connection));
-          });
-
-      // Connecting side.
-      q2.push(context->connect(listener->addr()));
-
-      // Run user specified functions.
-      std::thread t1([&] { f1(q1.pop()); });
-      std::thread t2([&] { f2(q2.pop()); });
-      t1.join();
-      t2.join();
-    }
-
-    context->join();
+    peers_->spawn(
+        [&] {
+          auto context = std::make_shared<tensorpipe::transport::uv::Context>();
+          context->setId("harness");
```

Comment:
As you're now creating two separate contexts in the client and server, could you give them different IDs? Like, `client_harness` and `server_harness`. Otherwise we won't be able to tell them apart in the logs